src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,189 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.Diagnostics
+{
+    internal class ConsoleWriter
+    {
+        // A typical out-of-proc function execution will generate 8 log lines.
+        // A single core container can potentially get around 1K RPS at the higher end, and a typical log line is around 300 bytes
+        // So in the extreme case, this is about 1 second of buffer and should be less than 3MB
+        private const int DefaultBufferSize = 8000;
+
+        private static readonly TimeSpan DefaultConsoleBufferTimeout = TimeSpan.FromSeconds(1);
+        private readonly Channel<string> _consoleBuffer;
+        private readonly TimeSpan _consoleBufferTimeout;
+        private readonly Action<Exception> _exceptionhandler;
+        private Task _consoleBufferReadLoop;
+        private Action<string> _writeEvent;
+
+        public ConsoleWriter(IEnvironment environment, Action<Exception> unhandledExceptionHandler)
+            : this(environment, unhandledExceptionHandler, consoleBufferTimeout: DefaultConsoleBufferTimeout, autoStart: true)
+        {
+        }
+
+        internal ConsoleWriter(IEnvironment environment, Action<Exception> exceptionHandler, TimeSpan consoleBufferTimeout, bool autoStart)
```

Comment:
this class is trying to perform 3 different modes: writer to buffer, write to `System.Console`, or no-op. Instead, can this be split into a contract and different implementations? An interface `IConsole`, three (or four for bounded vs unbounded channel) different implementations, then a helper method `public static IConsole CreateConsole(...args)` which will create the appropriate one based on supplied arguments?

Reply:
Along with the contract change, we could put this in the DI container and have it injected into `LinuxContainerEventGenerator`. I would also then recommend we adopt `IOptions` pattern for this (really going forward we should minimize any usage of `IEnvironment` or `IConfiguration` directly and opt for options patterns (only using `IEnvironment` or `IConfiguration` as part of options setup).

Reply:
> this class is trying to perform 3 different modes: writer to buffer, write to `System.Console`, or no-op. Instead, can this be split into a contract and different implementations? An interface `IConsoleWriter`, three (or four for bounded vs unbounded channel) different implementations, then a helper method `public static IConsole CreateConsole(...args)` which will create the appropriate one based on supplied arguments?

Doesn't it seem kind of over engineered to have 3 different types given that the no op and the direct writing to console are one liners? Its not like each of the methods is having to check various flags to figure out what mode its in. It also makes the fallback implementation (start using console.write directly once buffer is completed) more complicated once these modes are separated into different types.

The bounded vs unbounded behavior are the same so I'm not sure why we would want different classes for those, the only difference is whether we call `CreateBounded` or `CreateUnbounded`.

Reply:
> Along with the contract change, we could put this in the DI container and have it injected into `LinuxContainerEventGenerator`. I would also then recommend we adopt `IOptions` pattern for this. Really going forward we should minimize any usage of `IEnvironment` or `IConfiguration` directly and opt for options patterns (only using `IEnvironment` or `IConfiguration` as part of options setup).

Regarding registering ConsoleWriter - for types where there is literally just one consumer, I am not sure we want to clutter our DI registrations with them. For example, take a look at `WebHostMetricsLogger` - internally it uses a `MetricsEventManager` but because their lifetimes are tied together, the `MetricsEventManager` is not registered in the DI container and is instead constructed directly as part of the `WebHostMetricsLogger` constructor. I think this is better from an encapsulation perspective - otherwise the codepaths for registring services would continue to explode in size (they are already quite large).

Regarding configuration - I deliberately chose a style that is consistent with the related type - LinuxContainerEventGenerator, but I'll take another look.

Reply:
You don't have to separate into different types and use DI if you don't want to. However, I think while we only have 1 consumer right now, doesn't mean we won't have more in the future. I don't think someone in the future will take the time to move to DI when that 2nd consumer is added. Going the DI route also helps us follow the inversion-of-control pattern, and makes unit testing simpler. You could supply a test-implementation of `ConsoleWriter` - or the no-op implementation. You also don't need to think deeply about what values need to be in `IEnvironment`. 

As for the separate types - I like doing that for organization. I normally have the implementations as private nested classes. It helps me validate the behavior of each implementation (even the 1 liners), as I can consider the intended behavior of each implementation on their own.

But again - this is all up to you. I like it for code organization and testability.

Reply:
I think I found a sweet spot - I factored the implementation into `BufferedConsoleWriter` that just takes simple parameters and doesnt know anything about environment, added a ConsoleLoggingOptions type that LinuxContainerEventGenerator depends on, and pushed the logic for handling the environment into a ConsoleLoggingOptionsSetup type.